gammapy/estimators/energydependence.py

```diff
@@ -0,0 +1,264 @@
+        result_bkg_src = {}
+
+        result_bkg_src["Emin"] = self.energy_edges[:-1]
+        result_bkg_src["Emax"] = self.energy_edges[1:]
+        result_bkg_src["delta_ts"] = delta_ts_bkg_src
+        result_bkg_src["df"] = [df_bkg_src] * self.num_energy_bands
+        result_bkg_src["significance"] = [elem for elem in sigma_ts_bkg_src]
+
+        return result_bkg_src
+
+    def estimate_energy_dependence(self, datasets):
+        """Estimate the potential of energy-dependent morphology.
+
+        Parameters
+        ----------
+        datasets : `~gammapy.datasets.Datasets`
+            Input datasets to use.
+
+        Returns
+        -------
+        results : `dict`
+            Dictionary with results of the energy-dependence test. Entries are:
+            * "delta_ts" : difference in ts between fitting each energy band individually (sliced fit) and the joint fit
+            * "df" : the degrees of freedom between fitting each energy band individually (sliced fit) and the joint fit
+            * "result" : the results for the fitting each energy band individually (sliced fit) and the joint fit
+        """
+        model = datasets.models[self.source]
+
+        # Calculate the individually sliced components
+        slices_src = self._slice_datasets(datasets)
+        results_src = []
+        for sliced in slices_src:
+            results_src.append(self.fit.run(sliced))
+
+        results_src_total_stat = [result.total_stat for result in results_src]
+        free_x, free_y = np.shape(
+            [result.parameters.free_parameters.names for result in results_src]
+        )
+        df_src = free_x * free_y
+
+        # Calculate the joint fit
+        parameters = model.spatial_model.parameters.free_parameters.names
+        slice0 = slices_src[0]
+        for slice_j in slices_src[1:]:
+            for param in parameters:
+                setattr(
+                    slice_j.models[0].spatial_model,
+                    param,
+                    slice0.models[0].spatial_model.parameters[param],
+                )
+        result_joint = self.fit.run(slices_src)
+
+        # Compare fit of individual energy slices to the results with joint fit
+        delta_ts_joint = result_joint.total_stat - np.sum(results_src_total_stat)
+        df_joint = len(slices_src.parameters.free_parameters.names)
+        df = df_src - df_joint
+
+        # Prepare results dictionary
+        joint_values = [result_joint.parameters[param].value for param in parameters]
+        joint_errors = [result_joint.parameters[param].error for param in parameters]
+
+        parameter_values = np.empty((len(parameters), self.num_energy_bands))
+        parameter_errors = np.empty((len(parameters), self.num_energy_bands))
+        for i in range(self.num_energy_bands):
+            parameter_values[:, i] = [
+                results_src[i].parameters[param].value for param in parameters
```

Comment:
Is it always working? Shouldn't you rather access `results[i].models[self.source].free_parameters`?

Reply:
You could use `Parameters.value` to directly access all parameters values.

Reply:
Currently it is implemented through the parameters variable on line 193, such that the results will only include the spatial model!